From: Peter Oberparleiter <oberpar@linux.vnet.ibm.com>
Subject: cio: Fix missing subchannels after CHPID configure on
Patch-mainline: v3.14-rc5
Git-commit: 9955e8d15f53e53540aaed7bcef640142e65e900
References: bnc#866081, LTC#104808

Description:  cio: Fix missing subchannels after CHPID configure on
Symptom:      After configuring off all CHPIDs of an offline I/O device and
              subsequently configuring them on again using the Linux chchp
              command or by using the /sys/devices/css0/chp<CHPID>/configure
              sysfs attribute, the I/O device might not become available again.
Problem:      The Linux configure on function does not scan for subchannels
              that may have become available. This is normally not noticeable
              because the System z machine generates a Channel-Report-Word
              when the first CHPID of a subchannel is configured on, resulting
              in a scan of that subchannel. In case the first CHPID is not
              working properly though, configuring on any additional working
              CHPIDS will not have any effect on the availability of the
              affected subchannel.
Solution:     Fix this by ensuring that Linux performs a scan for new
              subchannels as part of the configure on function.
Reproduction: On a system where one channel-path is not working properly
              (for example because the associated cable has been pulled),
              choose an offline I/O device using that CHPID and at least
              one other working CHPID. Configure off all CHPIDs to that
              I/O device, then configure on the CHPIDs, starting with the
              faulty one. The I/O device will not become available unless
              other actions are taken (for example CHPID vary).

Signed-off-by: Peter Oberparleiter <oberpar@linux.vnet.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/cio/chsc.c |    1 +
 1 file changed, 1 insertion(+)

--- a/drivers/s390/cio/chsc.c
+++ b/drivers/s390/cio/chsc.c
@@ -533,6 +533,7 @@ void chsc_chp_online(struct chp_id chpid
 		css_wait_for_slow_path();
 		for_each_subchannel_staged(__s390_process_res_acc, NULL,
 					   &link);
+		css_schedule_reprobe();
 	}
 }
 
